# Introduction to pandas

Pandas is a useful python package for loading, manipulating, and exporting data tables (e.g., .csv files)

This notebook covers the basics of how to work with it, but only scratches the surface.

For more, please visit: https://pandas.pydata.org/docs/getting_started/tutorials.html


In [1]:
# Load pandas (and use the alias "pd" for convenience)

import pandas as pd

# Also, increase the number of columns that pandas will display by default (to avoid elipsis in the middle)

pd.set_option('display.max_columns', 1000)

# We'll also import numpy (with alias np) for additional functions, as well as the Counter object

import numpy as np
from collections import Counter



The most important type of object in pandas is the DataFrame. This is somewhat like a 2-dimensional array, but has additional funcationality.

When you load data from a file, pandas will put it into a DataFrame

There are separate functions for loading excel (.xls) and comma-separate values files (.csv)

Here, we will reuse the data files from the Tableau Assignment

In [2]:
df1 = pd.read_excel('movies_independent_part_1.xlsx')

# get the number of rows and columns

print(df1.shape)

(3201, 16)


We can use the `head` function to see the first few rows of the DataFrame

In [3]:
df1.head()

Title   US Gross Worldwide Gross  US DVD Sales  \
0              The Land Girls   146083.0        146083.0           NaN   
1      First Love, Last Rites    10876.0         10876.0           NaN   
2  I Married a Strange Person   203134.0        203134.0           NaN   
3        Let's Talk About Sex   373615.0        373615.0           NaN   
4                        Slam  1009819.0       1087521.0           NaN   

   Production Budget         Release Date MPAA Rating  Running Time (min)  \
0          8000000.0  1998-06-12 00:00:00           R                 NaN   
1           300000.0  1998-08-07 00:00:00           R                 NaN   
2           250000.0  1998-08-28 00:00:00         NaN                 NaN   
3           300000.0  1998-09-11 00:00:00         NaN                 NaN   
4          1000000.0  1998-10-09 00:00:00           R                 NaN   

  Distributor               Source Major Genre         Creative Type Director  \
0    Gramercy                  NaN         NaN                   NaN      NaN   
1      Strand                  NaN       Drama                   NaN      NaN   
2   Lionsgate                  NaN      Comedy                   NaN      NaN   
3   Fine Line                  NaN      Comedy                   NaN      NaN   
4     Trimark  Original Screenplay       Drama  Contemporary Fiction      NaN   

   Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  
0                     NaN          6.1      1071.0  
1                     NaN          6.9       207.0  
2                     NaN          6.8       865.0  
3                    13.0          NaN         NaN  
4                    62.0          3.4       165.0

## Missing values and data types

Note in the above table that pandas has replaced certain cells with "NaN" (not a number). This basically means that these are missing values (no data in that cell).

Although it is not immediately obvious, pandas has also made certain inferences about what type of data is in the table. We can get more information by checking the data types of each column using the `dtypes` attribute.

In [4]:
df1.dtypes

Title                      object
US Gross                   object
Worldwide Gross            object
US DVD Sales              float64
Production Budget         float64
Release Date               object
MPAA Rating                object
Running Time (min)        float64
Distributor                object
Source                     object
Major Genre                object
Creative Type              object
Director                   object
Rotten Tomatoes Rating    float64
IMDB Rating               float64
IMDB Votes                float64
dtype: object

Note that pandas has interpreted some columns as float (e.g.,"US DVD Sales"), and left the rest unspecified (the generic "object")

**Important**: note that some columns like "US Gross" have not been interpreted as floats, even though this seems like the right choice. We'll see why later...

For the numeric columns, we can get a summary of the data using the `describe` function

In [5]:
df1.describe()

US DVD Sales  Production Budget  Running Time (min)  \
count  5.640000e+02       3.200000e+03         1209.000000   
mean   3.490155e+07       3.106917e+07          110.193548   
std    4.589512e+07       3.558591e+07           20.171014   
min    6.184540e+05       2.180000e+02           46.000000   
25%    9.906211e+06       6.575000e+06           95.000000   
50%    2.033156e+07       2.000000e+07          107.000000   
75%    3.779422e+07       4.200000e+07          121.000000   
max    3.525821e+08       3.000000e+08          222.000000   

       Rotten Tomatoes Rating  IMDB Rating     IMDB Votes  
count             2321.000000  2988.000000    2988.000000  
mean                54.336924     6.283467   29908.644578  
std                 28.076593     1.252290   44937.582335  
min                  1.000000     1.400000      18.000000  
25%                 30.000000     5.600000    4828.500000  
50%                 55.000000     6.400000   15106.000000  
75%                 80.000000     7.200000   35810.500000  
max                100.000000     9.200000  519541.000000

Note that some each column has a different value for "count", because each column has a different number of missing values (NaNs). These NaNs are excluded from the calculations of mean, std, etc.

We can also see this directly using the `info` function:

In [6]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3201 entries, 0 to 3200
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Title                   3201 non-null   object 
 1   US Gross                3201 non-null   object 
 2   Worldwide Gross         3201 non-null   object 
 3   US DVD Sales            564 non-null    float64
 4   Production Budget       3200 non-null   float64
 5   Release Date            3201 non-null   object 
 6   MPAA Rating             2596 non-null   object 
 7   Running Time (min)      1209 non-null   float64
 8   Distributor             2969 non-null   object 
 9   Source                  2836 non-null   object 
 10  Major Genre             2926 non-null   object 
 11  Creative Type           2755 non-null   object 
 12  Director                1870 non-null   object 
 13  Rotten Tomatoes Rating  2321 non-null   float64
 14  IMDB Rating             2988 non-null   

## Indexing by row

By default, pandas will create an **index** when it reads in the data, such that each row has a unique identifier. To access the index, we can use the `index` attribute

In [7]:
df1.index

RangeIndex(start=0, stop=3201, step=1)

We can access a particular row using the index for that row using the `loc` attribute, with the index in square brackets

In [8]:
# get the row with index 0 (the first row)

df1.loc[0]

Title                          The Land Girls
US Gross                             146083.0
Worldwide Gross                      146083.0
US DVD Sales                              NaN
Production Budget                   8000000.0
Release Date              1998-06-12 00:00:00
MPAA Rating                                 R
Running Time (min)                        NaN
Distributor                          Gramercy
Source                                    NaN
Major Genre                               NaN
Creative Type                             NaN
Director                                  NaN
Rotten Tomatoes Rating                    NaN
IMDB Rating                               6.1
IMDB Votes                             1071.0
Name: 0, dtype: object

We can also access rows by the row number using the `iloc` feature. Here, they will be exactly the same since the index is just the integers starting at 0 (but this is not always the case).

In [9]:
df1.iloc[0]

Title                          The Land Girls
US Gross                             146083.0
Worldwide Gross                      146083.0
US DVD Sales                              NaN
Production Budget                   8000000.0
Release Date              1998-06-12 00:00:00
MPAA Rating                                 R
Running Time (min)                        NaN
Distributor                          Gramercy
Source                                    NaN
Major Genre                               NaN
Creative Type                             NaN
Director                                  NaN
Rotten Tomatoes Rating                    NaN
IMDB Rating                               6.1
IMDB Votes                             1071.0
Name: 0, dtype: object

You can also get multiple rows using the `X:Y` notation (with or without iloc)

In [10]:
df1.iloc[5:8]

Title   US Gross Worldwide Gross  US DVD Sales  \
5  Mississippi Mermaid    24551.0       2624551.0           NaN   
6            Following    44705.0         44705.0           NaN   
7              Foolish  6026908.0       6026908.0           NaN   

   Production Budget         Release Date MPAA Rating  Running Time (min)  \
5          1600000.0  1999-01-15 00:00:00         NaN                 NaN   
6             6000.0  1999-04-04 00:00:00           R                 NaN   
7          1600000.0  1999-04-09 00:00:00           R                 NaN   

  Distributor               Source Major Genre         Creative Type  \
5         MGM                  NaN         NaN                   NaN   
6   Zeitgeist                  NaN         NaN                   NaN   
7     Artisan  Original Screenplay      Comedy  Contemporary Fiction   

            Director  Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  
5                NaN                     NaN          NaN         NaN  
6  Christopher Nolan                     NaN          7.7     15133.0  
7                NaN                     NaN          3.8       353.0

In [11]:
df1[5:8]

Title   US Gross Worldwide Gross  US DVD Sales  \
5  Mississippi Mermaid    24551.0       2624551.0           NaN   
6            Following    44705.0         44705.0           NaN   
7              Foolish  6026908.0       6026908.0           NaN   

   Production Budget         Release Date MPAA Rating  Running Time (min)  \
5          1600000.0  1999-01-15 00:00:00         NaN                 NaN   
6             6000.0  1999-04-04 00:00:00           R                 NaN   
7          1600000.0  1999-04-09 00:00:00           R                 NaN   

  Distributor               Source Major Genre         Creative Type  \
5         MGM                  NaN         NaN                   NaN   
6   Zeitgeist                  NaN         NaN                   NaN   
7     Artisan  Original Screenplay      Comedy  Contemporary Fiction   

            Director  Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  
5                NaN                     NaN          NaN         NaN  
6  Christopher Nolan                     NaN          7.7     15133.0  
7                NaN                     NaN          3.8       353.0

Keep in mind the square brackets notation [ ] is **overloaded** so be careful when using it!

**Optional exercise: Try replacing the index (`df1.index`) with a 1-based index (rather than 0-based), and test out using `loc` vs `iloc`**

## Selecting by columns

You can get a list of the columns using the `columns` attribute. (Note that this returns the list of columns as an `Index` object)


In [12]:
df1.columns

Index(['Title', 'US Gross', 'Worldwide Gross', 'US DVD Sales',
       'Production Budget', 'Release Date', 'MPAA Rating',
       'Running Time (min)', 'Distributor', 'Source', 'Major Genre',
       'Creative Type', 'Director', 'Rotten Tomatoes Rating', 'IMDB Rating',
       'IMDB Votes'],
      dtype='object')

The easiest way to access columns is to use the column name (in square brackets). 

In [13]:
# Get the column for US Gross
df1['US Gross']

# This shows just the values for this column, along with the index

0         146083.0
1          10876.0
2         203134.0
3         373615.0
4        1009819.0
           ...    
3196    31452765.0
3197    33080084.0
3198    11989328.0
3199    45575336.0
3200    93828745.0
Name: US Gross, Length: 3201, dtype: object

You can also select multiple columns, by putting the names in a list

In [14]:
df1[['US Gross', 'Worldwide Gross', 'US DVD Sales']]

US Gross Worldwide Gross  US DVD Sales
0       146083.0        146083.0           NaN
1        10876.0         10876.0           NaN
2       203134.0        203134.0           NaN
3       373615.0        373615.0           NaN
4      1009819.0       1087521.0           NaN
...          ...             ...           ...
3196  31452765.0      36851125.0    21240321.0
3197  33080084.0      83080084.0    20983030.0
3198  11989328.0      12506188.0     6679409.0
3199  45575336.0     141475336.0           NaN
3200  93828745.0     233700000.0           NaN

[3201 rows x 3 columns]

The `loc` and `iloc` attributes can also be used to select rows and columsn simultaneously

In [15]:
# for loc, used named indices and columns
df1.loc[5:8, ['US Gross', 'Worldwide Gross']]

US Gross Worldwide Gross
5    24551.0       2624551.0
6    44705.0         44705.0
7  6026908.0       6026908.0
8  1641825.0       6341825.0

In [16]:
# for iloc, use numeric indices into the index and columns
df1.iloc[5:8, 1:3]

US Gross Worldwide Gross
5    24551.0       2624551.0
6    44705.0         44705.0
7  6026908.0       6026908.0

^ Note that the above behavioru can be confusing, especially if the index is non-numeric, or if the columns have numeriuc names. Usually it's best to keep things simple and make things as explicit as possible

## Summarizing non-numeric columns

How can we summarize the data in columns that have not been cast as numbers? Once again, the `describe` function can help.

In [17]:
# Summarize the data in the rating column:

df1['MPAA Rating'].describe()

count     2596
unique       7
top          R
freq      1194
Name: MPAA Rating, dtype: object

Here we can see that there are 2596 total (non-NaN) values, with 7 unique values, and that "R" is the most common value.

## Selecting subsets

One thing pandas is useful for is **filtering**. For example, we can select rows based on the values in certain columns.

This is a two step process (which we can combine into one) -- first make a boolean index, based on a set of criteria, and then use it to select rows

In [18]:
# Get the subset of films with an R rating

# First, create an index:
r_index = (df1['MPAA Rating'] == 'R')

# Pull out a subset based on this index
r_subset = df1[r_index]

# check the shape of the subset
print(r_subset.shape)

r_subset.head()

(1194, 16)


Title   US Gross Worldwide Gross  US DVD Sales  \
0          The Land Girls   146083.0        146083.0           NaN   
1  First Love, Last Rites    10876.0         10876.0           NaN   
4                    Slam  1009819.0       1087521.0           NaN   
6               Following    44705.0         44705.0           NaN   
7                 Foolish  6026908.0       6026908.0           NaN   

   Production Budget         Release Date MPAA Rating  Running Time (min)  \
0          8000000.0  1998-06-12 00:00:00           R                 NaN   
1           300000.0  1998-08-07 00:00:00           R                 NaN   
4          1000000.0  1998-10-09 00:00:00           R                 NaN   
6             6000.0  1999-04-04 00:00:00           R                 NaN   
7          1600000.0  1999-04-09 00:00:00           R                 NaN   

  Distributor               Source Major Genre         Creative Type  \
0    Gramercy                  NaN         NaN                   NaN   
1      Strand                  NaN       Drama                   NaN   
4     Trimark  Original Screenplay       Drama  Contemporary Fiction   
6   Zeitgeist                  NaN         NaN                   NaN   
7     Artisan  Original Screenplay      Comedy  Contemporary Fiction   

            Director  Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  
0                NaN                     NaN          6.1      1071.0  
1                NaN                     NaN          6.9       207.0  
4                NaN                    62.0          3.4       165.0  
6  Christopher Nolan                     NaN          7.7     15133.0  
7                NaN                     NaN          3.8       353.0

In [19]:
# Alternatively, do this in one step

r_subset = df1[(df1['MPAA Rating'] == 'R')]

print(r_subset.shape)

r_subset.head()

(1194, 16)


Title   US Gross Worldwide Gross  US DVD Sales  \
0          The Land Girls   146083.0        146083.0           NaN   
1  First Love, Last Rites    10876.0         10876.0           NaN   
4                    Slam  1009819.0       1087521.0           NaN   
6               Following    44705.0         44705.0           NaN   
7                 Foolish  6026908.0       6026908.0           NaN   

   Production Budget         Release Date MPAA Rating  Running Time (min)  \
0          8000000.0  1998-06-12 00:00:00           R                 NaN   
1           300000.0  1998-08-07 00:00:00           R                 NaN   
4          1000000.0  1998-10-09 00:00:00           R                 NaN   
6             6000.0  1999-04-04 00:00:00           R                 NaN   
7          1600000.0  1999-04-09 00:00:00           R                 NaN   

  Distributor               Source Major Genre         Creative Type  \
0    Gramercy                  NaN         NaN                   NaN   
1      Strand                  NaN       Drama                   NaN   
4     Trimark  Original Screenplay       Drama  Contemporary Fiction   
6   Zeitgeist                  NaN         NaN                   NaN   
7     Artisan  Original Screenplay      Comedy  Contemporary Fiction   

            Director  Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  
0                NaN                     NaN          6.1      1071.0  
1                NaN                     NaN          6.9       207.0  
4                NaN                    62.0          3.4       165.0  
6  Christopher Nolan                     NaN          7.7     15133.0  
7                NaN                     NaN          3.8       353.0

We can also combine two criteria into one selection; we can use the `&` operater so select rows that satisfy **both** criteria

In [20]:
# Select columns based on both rating and budget

subset2 = df1[(df1['MPAA Rating'] == 'R') & (df1['Production Budget'] > 8000000)]

print(subset2.shape)
subset2.head()

(821, 16)


Title    US Gross Worldwide Gross  \
8                                      Pirates   1641825.0       6341825.0   
20                              Twelve Monkeys  57141459.0     168841459.0   
29  Three Kingdoms: Resurrection of the Dragon         0.0      22139590.0   
52                                       Alive  36299670.0      36299670.0   
61                              Apocalypse Now  78800000.0      78800000.0   

    US DVD Sales  Production Budget         Release Date MPAA Rating  \
8            NaN         40000000.0  1986-07-01 00:00:00           R   
20           NaN         29000000.0  1995-12-27 00:00:00           R   
29           NaN         20000000.0  2008-04-03 00:00:00           R   
52           NaN         32000000.0  1993-01-15 00:00:00           R   
61     3479242.0         31500000.0  1979-08-15 00:00:00           R   

    Running Time (min)           Distributor                     Source  \
8                  NaN                   NaN                        NaN   
20                 NaN             Universal        Based on Short Film   
29                 NaN                   NaN  Based on Book/Short Story   
52                 NaN  Walt Disney Pictures  Based on Book/Short Story   
61                 NaN                   MGM  Based on Book/Short Story   

   Major Genre       Creative Type              Director  \
8          NaN                 NaN        Roman Polanski   
20       Drama     Science Fiction         Terry Gilliam   
29      Action  Historical Fiction                   NaN   
52   Adventure       Dramatization        Frank Marshall   
61      Action  Historical Fiction  Francis Ford Coppola   

    Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  
8                     25.0          5.8      3275.0  
20                     NaN          8.1    169858.0  
29                     NaN          NaN         NaN  
52                    71.0          3.2       124.0  
61                    98.0          8.6    173141.0

### Optional Exercise: Get the films that are Dramas with Rotten Tomatoes Rating > 90 and IMDB Rating >= 8
Hint: this should return 33 films

## Extracting data from the DataFrame

Sometimes it's easiset to pull data out of the DataFrame and work with it externally.

To extract the values in a column into an array, we can use the `values` attribute.

We can also cast this into a `list`, or other data types like `set`

For example, let's see what the various MPAA ratings are

In [21]:
# First, just get the column
ratings_column = df1['MPAA Rating']
print(type(ratings_column))

# Compare to getting the actual values
ratings = df1['MPAA Rating'].values
print(type(ratings))

<class 'pandas.core.series.Series'>
<class 'numpy.ndarray'>


In [22]:
# convert this to a set to see what types of ratings there are
set(ratings)

{'G', 'NC-17', 'Not Rated', 'Open', 'PG', 'PG-13', 'R', nan}

In [23]:
# Use a Counter to see how common each one is
rating_counter = Counter(df1['MPAA Rating'].values)

for rating, count in rating_counter.most_common():
    print(rating, count)

R 1194
PG-13 865
nan 605
PG 354
Not Rated 94
G 79
NC-17 8
Open 2


In [24]:
# Let's see what the Open rating is
df1[df1['MPAA Rating'] == 'Open']

Title   US Gross Worldwide Gross  US DVD Sales  \
2171               L.I.E.  1138836.0       1138836.0           NaN   
2654  Requiem for a Dream  3635482.0       7390108.0           NaN   

      Production Budget         Release Date MPAA Rating  Running Time (min)  \
2171           700000.0  2001-09-07 00:00:00        Open                 NaN   
2654          4500000.0  2000-10-06 00:00:00        Open                 NaN   

       Distributor                     Source Major Genre  \
2171  Lot 47 Films        Original Screenplay       Drama   
2654       Artisan  Based on Book/Short Story       Drama   

             Creative Type          Director  Rotten Tomatoes Rating  \
2171  Contemporary Fiction               NaN                    83.0   
2654  Contemporary Fiction  Darren Aronofsky                    78.0   

      IMDB Rating  IMDB Votes  
2171          7.2      5122.0  
2654          8.5    185226.0

**Optional Exercise: Find the top 3 (most common) distributors**

## Investigating data types

Let's see why US Gross didn't get converted to a float. In this case, we'll see that pandas interpreted some values as floats, and others as strings.

In [25]:
# get the US Gross values
gross_values = df1['US Gross'].values

gross_datatypes = [type(v) for v in gross_values]
print(set(gross_datatypes))

{<class 'float'>, <class 'str'>}


In [26]:
# Let's look at the values that are still strings

print(set([v for v in gross_values if type(v) == str]))

{'Unknown'}


As we can see, some rows have MPAA rating listed as "Unknown"

Let's replace those values with NaNs

## Modifying values and Adding to a DataFrame

To add a new columns (or overwrite an old one), we can simply assign an array or list using the column name. Note that the length of the list should be the same as the number of rows in the DataFrame

In [27]:
# get the US Gross column
gross_column = df1['US Gross']

# Replace unknown values with NaNs
# Here we'll use the NaN value from numpy (np.nan)
gross_column_clean = gross_column.replace(to_replace='Unknown', value=np.nan, inplace=False)

# Add a new column with the cleaned up values
df1['US Gross Clean'] = gross_column_clean

In [28]:
# get the cleaned up US Gross values
gross_values = df1['US Gross Clean'].values

gross_datatypes = [type(v) for v in gross_values]
print(set(gross_datatypes))

{<class 'numpy.float64'>}


In [29]:
# Verify to make sure things got replaced properly

# Select the rows that had US Gross = Unknown
unknown_gross_subset = df1[(df1['US Gross'] == 'Unknown')]

# Show a few columns of this subset:
unknown_gross_subset[['Title', 'US Gross', 'US Gross Clean']]

# Indeed, they are now NaNs in the new column

Title US Gross  US Gross Clean
118           Bananas  Unknown             NaN
254   Damnation Alley  Unknown             NaN
266   Death Race 2000  Unknown             NaN
404     Hell's Angels  Unknown             NaN
467       Intolerance  Unknown             NaN
1025         Waterloo  Unknown             NaN
1028            Wings  Unknown             NaN

In [30]:
# Check the data type of the new column
df1.dtypes

# Indeed, it is now a float64

Title                      object
US Gross                   object
Worldwide Gross            object
US DVD Sales              float64
Production Budget         float64
Release Date               object
MPAA Rating                object
Running Time (min)        float64
Distributor                object
Source                     object
Major Genre                object
Creative Type              object
Director                   object
Rotten Tomatoes Rating    float64
IMDB Rating               float64
IMDB Votes                float64
US Gross Clean            float64
dtype: object

## Global replacement

Alternatively, we can use the `replace` function to replace values globally. Note that this method should be used with care!



In [31]:
# First, check to see how many times the value "Unknown" occurs in each column:
(df1 == 'Unknown').sum(axis=0)

Title                     1
US Gross                  7
Worldwide Gross           7
US DVD Sales              0
Production Budget         0
Release Date              0
MPAA Rating               0
Running Time (min)        0
Distributor               0
Source                    0
Major Genre               0
Creative Type             0
Director                  0
Rotten Tomatoes Rating    0
IMDB Rating               0
IMDB Votes                0
US Gross Clean            0
dtype: int64

In [32]:
# It looks quite rare, so we can simply replace all "Unknown" values with NaN  (inplace controls if this generates a new copy or not)
df1_clean = df1.replace(to_replace='Unknown', value=np.nan, inplace=False)

# Check that all Unknown values are all gone
(df1_clean == 'Unknown').sum()

Title                     0
US Gross                  0
Worldwide Gross           0
US DVD Sales              0
Production Budget         0
Release Date              0
MPAA Rating               0
Running Time (min)        0
Distributor               0
Source                    0
Major Genre               0
Creative Type             0
Director                  0
Rotten Tomatoes Rating    0
IMDB Rating               0
IMDB Votes                0
US Gross Clean            0
dtype: int64

## Date conversion

In [33]:
# Let's get the year of each film

# First, check the data types in the Release Date column

date_values = df1['Release Date'].values

date_datatypes = [type(v) for v in date_values]
print(set(date_datatypes))

{<class 'str'>, <class 'float'>, <class 'datetime.datetime'>}


In [34]:
# import the datetime package to deal with those
import datetime

In [35]:
# Let's look at the values that are strings
print("str dates")
print([v for v in date_values if type(v) == str])
print()

# and those that are floast (print the first ten)
print("float dates")
print([v for v in date_values if type(v) == float][:10])
print()

# and those that are datetimes (print the first ten)
print("float dates")
print([v for v in date_values if type(v) == datetime.datetime][:10])

str dates
['TBD', 'TBD', 'TBD', 'TBD', 'TBD', 'TBD', 'TBD']

float dates
[1963.0, 1938.0, 1987.0, 1960.0, 1975.0, 1973.0, 1991.0, 1976.0, 1929.0, 1940.0]

float dates
[datetime.datetime(1998, 6, 12, 0, 0), datetime.datetime(1998, 8, 7, 0, 0), datetime.datetime(1998, 8, 28, 0, 0), datetime.datetime(1998, 9, 11, 0, 0), datetime.datetime(1998, 10, 9, 0, 0), datetime.datetime(1999, 1, 15, 0, 0), datetime.datetime(1999, 4, 4, 0, 0), datetime.datetime(1999, 4, 9, 0, 0), datetime.datetime(1986, 7, 1, 0, 0), datetime.datetime(1946, 12, 31, 0, 0)]


We can see from this that some have been interpreted as floats (just a year), whereas others have been interpreted as richer date objects (year, month, day, hour, minute)

In [36]:
# Extract the year depending on the data time

# create an empty list to store the years
years = []

# process each date value in turn
for value in date_values:
    # if it's a string (TBD), set year to np.nan
    if type(value) == str:
        years.append(np.nan)
    # If it's a float (just a year), convert it to an int
    elif type(value) == float:
        years.append(int(value))
    # otherwise, get the year from the datetime object
    else:
        years.append(int(value.year))

# finally, add a new column to the dataset to hold the years
df1['Year'] = years

df1.head()

Title   US Gross Worldwide Gross  US DVD Sales  \
0              The Land Girls   146083.0        146083.0           NaN   
1      First Love, Last Rites    10876.0         10876.0           NaN   
2  I Married a Strange Person   203134.0        203134.0           NaN   
3        Let's Talk About Sex   373615.0        373615.0           NaN   
4                        Slam  1009819.0       1087521.0           NaN   

   Production Budget         Release Date MPAA Rating  Running Time (min)  \
0          8000000.0  1998-06-12 00:00:00           R                 NaN   
1           300000.0  1998-08-07 00:00:00           R                 NaN   
2           250000.0  1998-08-28 00:00:00         NaN                 NaN   
3           300000.0  1998-09-11 00:00:00         NaN                 NaN   
4          1000000.0  1998-10-09 00:00:00           R                 NaN   

  Distributor               Source Major Genre         Creative Type Director  \
0    Gramercy                  NaN         NaN                   NaN      NaN   
1      Strand                  NaN       Drama                   NaN      NaN   
2   Lionsgate                  NaN      Comedy                   NaN      NaN   
3   Fine Line                  NaN      Comedy                   NaN      NaN   
4     Trimark  Original Screenplay       Drama  Contemporary Fiction      NaN   

   Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  US Gross Clean    Year  
0                     NaN          6.1      1071.0        146083.0  1998.0  
1                     NaN          6.9       207.0         10876.0  1998.0  
2                     NaN          6.8       865.0        203134.0  1998.0  
3                    13.0          NaN         NaN        373615.0  1998.0  
4                    62.0          3.4       165.0       1009819.0  1998.0

In [37]:
df1.describe()

US DVD Sales  Production Budget  Running Time (min)  \
count  5.640000e+02       3.200000e+03         1209.000000   
mean   3.490155e+07       3.106917e+07          110.193548   
std    4.589512e+07       3.558591e+07           20.171014   
min    6.184540e+05       2.180000e+02           46.000000   
25%    9.906211e+06       6.575000e+06           95.000000   
50%    2.033156e+07       2.000000e+07          107.000000   
75%    3.779422e+07       4.200000e+07          121.000000   
max    3.525821e+08       3.000000e+08          222.000000   

       Rotten Tomatoes Rating  IMDB Rating     IMDB Votes  US Gross Clean  \
count             2321.000000  2988.000000    2988.000000    3.194000e+03   
mean                54.336924     6.283467   29908.644578    4.400209e+07   
std                 28.076593     1.252290   44937.582335    6.255531e+07   
min                  1.000000     1.400000      18.000000    0.000000e+00   
25%                 30.000000     5.600000    4828.500000    5.493221e+06   
50%                 55.000000     6.400000   15106.000000    2.201947e+07   
75%                 80.000000     7.200000   35810.500000    5.609176e+07   
max                100.000000     9.200000  519541.000000    7.601676e+08   

              Year  
count  3194.000000  
mean   1997.992486  
std      12.158235  
min    1929.000000  
25%    1996.000000  
50%    2001.000000  
75%    2005.000000  
max    2027.000000

**Optional Exercise: Create a column for Month**

## Dealing with NaNs

A common issue is how to deal with missing values.

One approach is to simply drop those rows for some column.

The `isna` or `notna` functions can be used to generate an index, which we can use to select

In [38]:
# get the non-NaN US DVD Sales
temp_index = df1['US DVD Sales'].notna()

dvd_subset = df1[temp_index]
print(dvd_subset.shape)
dvd_subset.head()

(564, 18)


Title     US Gross Worldwide Gross  \
61                             Apocalypse Now   78800000.0      78800000.0   
289  Star Wars Ep. V: The Empire Strikes Back  290271960.0     534171960.0   
453                     It's a Wonderful Life    6600000.0       6600000.0   
640      Indiana Jones and the Temple of Doom  179880271.0     333080271.0   
641        Indiana Jones and the Last Crusade  197171806.0     474171806.0   

     US DVD Sales  Production Budget         Release Date MPAA Rating  \
61      3479242.0         31500000.0  1979-08-15 00:00:00           R   
289    10027926.0         23000000.0  1980-05-21 00:00:00          PG   
453    19339789.0          3180000.0               1946.0         NaN   
640    18998388.0         28000000.0  1984-05-23 00:00:00         NaN   
641    18740425.0         48000000.0  1989-05-24 00:00:00         NaN   

     Running Time (min)         Distributor                     Source  \
61                  NaN                 MGM  Based on Book/Short Story   
289                 NaN    20th Century Fox        Original Screenplay   
453                 NaN                 NaN                        NaN   
640                 NaN  Paramount Pictures        Original Screenplay   
641                 NaN  Paramount Pictures        Original Screenplay   

    Major Genre       Creative Type              Director  \
61       Action  Historical Fiction  Francis Ford Coppola   
289   Adventure     Science Fiction                   NaN   
453         NaN                 NaN           Frank Capra   
640   Adventure  Historical Fiction      Steven Spielberg   
641   Adventure  Historical Fiction      Steven Spielberg   

     Rotten Tomatoes Rating  IMDB Rating  IMDB Votes  US Gross Clean    Year  
61                     98.0          8.6    173141.0      78800000.0  1979.0  
289                     NaN          NaN         NaN     290271960.0  1980.0  
453                    94.0          8.7    101499.0       6600000.0  1946.0  
640                    85.0          7.5    110761.0     179880271.0  1984.0  
641                    89.0          8.3    171572.0     197171806.0  1989.0

In [39]:
# Alterantively, we can drop rows that are ALL NaNs or have ANY NaNs using the dropna function
df1_no_nans = df1.dropna(how='any')

print(df1_no_nans.shape)
df1_no_nans.head()

# Note that this only leaves us with 174 rows! (that don't have any NaNs in any column)

(174, 18)


Title     US Gross Worldwide Gross  US DVD Sales  \
1064     12 Rounds   12234694.0      18184083.0     8283859.0   
1074        2012.0  166112167.0     766812167.0    50736023.0   
1090         300.0  210614939.0     456068181.0   261252400.0   
1095  3:10 to Yuma   53606916.0      69791889.0    51359371.0   
1107    88 Minutes   16930884.0      32955399.0    11385055.0   

      Production Budget         Release Date MPAA Rating  Running Time (min)  \
1064         20000000.0  2009-03-27 00:00:00       PG-13               108.0   
1074        200000000.0  2009-11-13 00:00:00       PG-13               158.0   
1090         60000000.0  2007-03-09 00:00:00           R               117.0   
1095         48000000.0  2007-09-02 00:00:00           R               117.0   
1107         30000000.0  2008-04-18 00:00:00           R               106.0   

           Distributor                        Source        Major Genre  \
1064  20th Century Fox           Original Screenplay             Action   
1074     Sony Pictures           Original Screenplay             Action   
1090      Warner Bros.  Based on Comic/Graphic Novel             Action   
1095         Lionsgate                        Remake            Western   
1107     Sony Pictures           Original Screenplay  Thriller/Suspense   

             Creative Type         Director  Rotten Tomatoes Rating  \
1064  Contemporary Fiction     Renny Harlin                    28.0   
1074       Science Fiction  Roland Emmerich                    39.0   
1090    Historical Fiction      Zack Snyder                    60.0   
1095    Historical Fiction    James Mangold                    89.0   
1107  Contemporary Fiction        Jon Avnet                     5.0   

      IMDB Rating  IMDB Votes  US Gross Clean    Year  
1064          5.4      8914.0      12234694.0  2009.0  
1074          6.2       396.0     166112167.0  2009.0  
1090          7.8    235508.0     210614939.0  2007.0  
1095          7.9     98355.0      53606916.0  2007.0  
1107          5.9     31205.0      16930884.0  2008.0

In [40]:
# Or, we can use the fillna() function to replace values in various ways

# In this case, let's replace US Gross Clean NaNs with the US Gross median

us_gross_median = df1['US Gross Clean'].median()
print(us_gross_median)

gross_replaced = df1['US Gross Clean'].fillna(us_gross_median, inplace=False)

df1['US Gross Clean Replaced'] = gross_replaced

# Note that the means of these columns will now be slightly different
df1[['US Gross Clean', 'US Gross Clean Replaced']].describe()

22019465.5


US Gross Clean  US Gross Clean Replaced
count    3.194000e+03             3.201000e+03
mean     4.400209e+07             4.395401e+07
std      6.255531e+07             6.249529e+07
min      0.000000e+00             0.000000e+00
25%      5.493221e+06             5.501940e+06
50%      2.201947e+07             2.201947e+07
75%      5.609176e+07             5.606855e+07
max      7.601676e+08             7.601676e+08

**Note:** More sophisticated approaches would try to **infer** missing values by using all available data, but that is beyond the scope of this course

**Optional exercise: Try to infer missing Rotten Tomatoes Ratings by using a model based on US Gross**

## Loading csv files

We can also load csv files; sometimes here it helps to exert more control

In [41]:
# Load the CSV file
df2 = pd.read_csv('movies_independent_part_2.csv')
df2.head()

Unnamed: 0  year       imdb             title             test clean_test  \
0           0  2013  tt1711425     21 &amp; Over           notalk     notalk   
1           1  2012  tt1343727          Dredd 3D      ok-disagree         ok   
2           2  2013  tt2024544  12 Years a Slave  notalk-disagree     notalk   
3           3  2013  tt1272878            2 Guns           notalk     notalk   
4           4  2013  tt0453562                42              men        men   

  binary    budget    domgross     intgross      code  budget_2013$  \
0   FAIL  13000000  25682380.0   42195766.0  2013FAIL      13000000   
1   PASS  45000000  13414714.0   40868994.0  2012PASS      45658735   
2   FAIL  20000000  53107035.0  158607035.0  2013FAIL      20000000   
3   FAIL  61000000  75612460.0  132493015.0  2013FAIL      61000000   
4   FAIL  40000000  95020213.0   95020213.0  2013FAIL      40000000   

   domgross_2013$  intgross_2013$  period code  decade code  \
0      25682380.0      42195766.0          1.0          1.0   
1      13611086.0      41467257.0          1.0          1.0   
2      53107035.0     158607035.0          1.0          1.0   
3      75612460.0     132493015.0          1.0          1.0   
4      95020213.0      95020213.0          1.0          1.0   

            director director_gender      genre  rating         country  \
0          Jon Lucas            male     Comedy     5.9   United States   
1        Pete Travis            male     Action     7.1  United Kingdom   
2      Steve McQueen            male  Biography     8.1   United States   
3  Baltasar Kormákur            male     Action     6.7   United States   
4    Brian Helgeland            male  Biography     7.5   United States   

  language  
0  English  
1  English  
2  English  
3  English  
4  English

In [42]:
# We can see that the first column was already an index

# Let's reload it and tell it to use that column as the index

# Include other options as demonstration (e.g., is there a header? how are the columns separated? etc.)

df2 = pd.read_csv('movies_independent_part_2.csv', index_col=0, header=0, sep=',')
df2.head()

year       imdb             title             test clean_test binary  \
0  2013  tt1711425     21 &amp; Over           notalk     notalk   FAIL   
1  2012  tt1343727          Dredd 3D      ok-disagree         ok   PASS   
2  2013  tt2024544  12 Years a Slave  notalk-disagree     notalk   FAIL   
3  2013  tt1272878            2 Guns           notalk     notalk   FAIL   
4  2013  tt0453562                42              men        men   FAIL   

     budget    domgross     intgross      code  budget_2013$  domgross_2013$  \
0  13000000  25682380.0   42195766.0  2013FAIL      13000000      25682380.0   
1  45000000  13414714.0   40868994.0  2012PASS      45658735      13611086.0   
2  20000000  53107035.0  158607035.0  2013FAIL      20000000      53107035.0   
3  61000000  75612460.0  132493015.0  2013FAIL      61000000      75612460.0   
4  40000000  95020213.0   95020213.0  2013FAIL      40000000      95020213.0   

   intgross_2013$  period code  decade code           director  \
0      42195766.0          1.0          1.0          Jon Lucas   
1      41467257.0          1.0          1.0        Pete Travis   
2     158607035.0          1.0          1.0      Steve McQueen   
3     132493015.0          1.0          1.0  Baltasar Kormákur   
4      95020213.0          1.0          1.0    Brian Helgeland   

  director_gender      genre  rating         country language  
0            male     Comedy     5.9   United States  English  
1            male     Action     7.1  United Kingdom  English  
2            male  Biography     8.1   United States  English  
3            male     Action     6.7   United States  English  
4            male  Biography     7.5   United States  English